packages/packagers/js/src/ScopeHoistingPackager.js

```diff
@@ -0,0 +1,1049 @@
+      }
+    });
+
+    this.assetOutputs = new Map(await queue.run());
+  }
+
+  buildExportedSymbols() {
+    if (this.isAsyncBundle || this.bundle.env.outputFormat !== 'esmodule') {
+      return;
+    }
+
+    let entry = this.bundle.getMainEntry();
+    if (entry) {
+      for (let {exportAs, symbol} of this.bundleGraph.getExportedSymbols(
+        entry,
+      )) {
+        if (typeof symbol === 'string') {
+          let symbols = this.exportedSymbols.get(
+            symbol === '*' ? nullthrows(entry.symbols.get('*')?.local) : symbol,
+          );
+
+          let local = symbol;
+          if (symbols) {
+            local = symbols[0].local;
+          } else {
+            symbols = [];
+            this.exportedSymbols.set(symbol, symbols);
+          }
+
+          if (exportAs === '*') {
+            exportAs = 'default';
+          }
+
+          symbols.push({exportAs, local});
+        } else if (symbol === null) {
+          // TODO `meta.exportsIdentifier[exportSymbol]` should be exported
+          // let relativePath = relative(options.projectRoot, asset.filePath);
+          // throw getThrowableDiagnosticForNode(
+          //   md`${relativePath} couldn't be statically analyzed when importing '${exportSymbol}'`,
+          //   entry.filePath,
+          //   loc,
+          // );
+        } else if (symbol !== false) {
+          // let relativePath = relative(options.projectRoot, asset.filePath);
+          // throw getThrowableDiagnosticForNode(
+          //   md`${relativePath} does not export '${exportSymbol}'`,
+          //   entry.filePath,
+          //   loc,
+          // );
+        }
+      }
+    }
+  }
+
+  getTopLevelName(name: string): string {
+    name = name.replace(NON_ID_CONTINUE_RE, '');
+    if (!ID_START_RE.test(name) || this.globalNames.has(name)) {
+      name = '_' + name;
+    }
+
+    let count = this.topLevelNames.get(name);
+    if (count == null) {
+      this.topLevelNames.set(name, 1);
+      return name;
+    }
+
+    this.topLevelNames.set(name, count + 1);
+    return name + count;
+  }
+
+  visitAsset(asset: Asset): [string, ?SourceMap, number] {
+    invariant(!this.seenAssets.has(asset.id), 'Already visited asset');
+    this.seenAssets.add(asset.id);
+
+    let {code, map} = nullthrows(this.assetOutputs.get(asset.id));
+    return this.buildAsset(asset, code, map);
+  }
+
+  buildAsset(
+    asset: Asset,
+    code: string,
+    map: ?Buffer,
+  ): [string, ?SourceMap, number] {
+    let shouldWrap = this.wrappedAssets.has(asset.id);
+    let deps = this.bundleGraph.getDependencies(asset);
+
+    let sourceMap = this.bundle.env.sourceMap
+      ? new SourceMap(this.options.projectRoot)
+      : null;
+    if (sourceMap && map) {
+      sourceMap?.addBufferMappings(map);
+    }
+
+    // If this asset is skipped, just add dependencies and not the asset's content.
+    if (this.shouldSkipAsset(asset)) {
+      let depCode = '';
+      let lineCount = 0;
+      // TODO: order?
```

Comment:
I just meant it's relying on the order in the graph rather than the order of the `import` statements left in the code. But haven't run into any issues so far...